velox/common/memory/MallocAllocator.h

```diff
@@ -90,11 +90,35 @@ class MallocAllocator : public MemoryAllocator {
 
   void freeContiguousImpl(ContiguousAllocation& allocation);
 
+  /// Increment current usage and check current allocator consistency to make
+  /// sure current usage does not go above 'capacity_'. If it goes above
+  /// 'capacity_', the increment will not be applied. Returns true if within
+  /// capacity, false otherwise.
+  ///
+  /// NOTE: This method should always be called BEFORE actual allocation.
+  bool incrementUsage(int64_t bytes);
+
+  /// Decrement current usage and check current allocator consistency to make
+  /// sure current usage does not go below 0. Throws if usage goes below 0.
+  ///
+  /// NOTE: This method should always be called AFTER actual free.
+  void decrementUsage(int64_t bytes);
+
   const Kind kind_;
 
+  /// Capacity in bytes. Total allocation byte is not allowed to exceed this
+  /// value. Setting this to 0 means no capacity enforcement.
+  const size_t capacity_;
+
+  /// Current total allocated bytes by this 'MallocAllocator'.
+  std::atomic<int64_t> allocatedBytes_{0};
```

Comment:
Can you add to check allocatedBytes_ is zero in dtor. Also consider to move allocatedBytes_ to base class and use it for allocated bytes for both allocators. We don't need the separate counter for allocated pages. We could consider this code consolidation in followup. Thanks!